src/riak_cs_gc_manager.erl

```diff
@@ -0,0 +1,294 @@
+%%% API
+%%%===================================================================
+
+start_batch(Options) ->
+    gen_fsm:sync_send_event(?SERVER, {start, Options}, infinity).
+
+stop_batch() ->
+    gen_fsm:sync_send_event(?SERVER, stop, infinity).
+
+pause_batch() ->
+    gen_fsm:sync_send_event(?SERVER, pause, infinity).
+
+resume_batch() ->
+    gen_fsm:sync_send_event(?SERVER, resume, infinity).
+
+status() ->
+    gen_fsm:sync_send_all_state_event(?SERVER, status, infinity).
+
+%% @doc Adjust the interval at which the daemon attempts to perform
+%% a garbage collection sweep. Setting the interval to a value of
+%% `infinity' effectively disable garbage collection. The daemon still
+%% runs, but does not carry out any file deletion.
+-spec set_interval(term()) -> ok | {error, term()}.
+set_interval(Interval) when is_integer(Interval)
+                            orelse Interval =:= infinity ->
+    gen_fsm:sync_send_all_state_event(?SERVER, {set_interval, Interval}, infinity).
+
+
+start_link() ->
+    gen_fsm:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+-ifdef(TEST).
+test_link() ->
+    gen_fsm:start_link({local, ?SERVER}, ?MODULE, [testing], []).
+
+-endif.
+
+
+finished(Report) ->
+    gen_fsm:sync_send_event(?SERVER, {finished, Report}, infinity).
+
+%%%===================================================================
+%%% gen_fsm callbacks
+%%%===================================================================
+
+init([]) ->
+    InitialDelay = riak_cs_gc:initial_gc_delay(),
+    _ = case riak_cs_gc:gc_interval() of
+            infinity ->
+                ok;
+            Leeway when is_integer(Leeway) ->
+                _TimerRef = erlang:send_after(InitialDelay * 1000, self(), {start, []})
+        end,
+    {ok, idle, #gc_manager_state{}};
+init([testing]) ->
+    {ok, idle, #gc_manager_state{}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc All gen_fsm:send_event/2 call should be ignored
+idle(_Event, State) ->
+    {next_state, idle, State}.
+running(_Event, State) ->
+    {next_state, running, State}.
+paused_finished(_Event, State) ->
+    {next_state, paused_finished, State}.
+paused(_Event, State) ->
+    {next_state, paused, State}.
+
+%% @private
+%% @doc
+idle({start, Options}, _From, State) ->
+    MaxWorkers =  riak_cs_gc:gc_max_workers(),
+    %% StartKey = proplists:get_value(start, Options,
+    %%                                riak_cs_gc:epoch_start()),
+    BatchStart = riak_cs_gc:timestamp(),
+    %% EndKey = proplists:get_value('end', Options, BatchStart),
+    Leeway = proplists:get_value(leeway, Options,
+                                 riak_cs_gc:leeway_seconds()),
+                                 
+    %% set many items to GCDState here
+    GCDState = #gc_d_state{
+                  batch_start=BatchStart,
+                  leeway=Leeway,
+                  max_workers=MaxWorkers},
+
+    case riak_cs_gc_d:start_link(GCDState) of
```

Comment:
This call may be blocked by 2i fetch and it may be long depending on cluster / nodes status.